File: pages/BasePage.js
'use strict';

var driver;
var Promise = require('selenium-webdriver').promise;
var Until = require('selenium-webdriver').until;

function BasePage(driver) {
  this.driver = driver;
}

BasePage.prototype.visit = function(url) {
  this.driver.get(url);
};

BasePage.prototype.find = function(locator) {
  return this.driver.findElement(locator);
};

BasePage.prototype.click = function(locator) {
  this.find(locator).click();
};

BasePage.prototype.type = function(locator, inputText) {
  this.find(locator).sendKeys(inputText);
};

BasePage.prototype.isDisplayed = function(locator) {
  var defer = Promise.defer();
  this.find(locator).isDisplayed().then(function(isDisplayed) {
    defer.fulfill(isDisplayed);
  }, function(error) {
    if (error.name === 'NoSuchElementError') {
      defer.fulfill(false);
    } else {
      defer.reject(error);
    }
  });
  return defer.promise;
};

BasePage.prototype.waitForIsDisplayed = function(locator, timeout) {
  var defer = Promise.defer();
  var driver = this.driver;
  driver.wait(Until.elementLocated(locator), timeout).then(function() {
    var element = driver.findElement(locator);
    driver.wait(Until.elementIsVisible(element), timeout).then(function() {
      defer.fulfill(true);
    }, function(error) {
      if (error.name === 'NoSuchElementError') {
        defer.fulfill(false);
      } else {
        defer.reject(error);
      }
    });
  });
  return defer.promise;
};

module.exports = BasePage;

File: code-examples/ruby/09/02/base_page.rb
require 'selenium-webdriver'

class BasePage

  def initialize(driver)
    @driver = driver
  end

  def visit(url_path)
    @driver.get ENV['base_url'] + url_path
  end

  def find(locator)
    @driver.find_element locator
  end

  def type(text, locator)
    find(locator).send_keys text
  end

  def submit(locator)
    find(locator).submit
  end

  def is_displayed?(locator)
    begin
      find(locator).displayed?
    rescue Selenium::WebDriver::Error::NoSuchElementError
      false
    end
  end

end
